kernel/preemption/src/lib.rs

```diff
@@ -9,26 +9,30 @@
 extern crate alloc;
 
 use core::sync::atomic::{AtomicU8, Ordering};
+use cpu::CpuId;
 
-/// Theseus uses a `u8` to hold each CPU core's ID (apic ID),
-/// so the maximum number of cores it supports is `u8::MAX` (256).
-const MAX_CPU_CORES: usize = u8::MAX as usize;
+/// Theseus previously used a `u8` to hold each CPU core's ID
+/// so the maximum number of cores it supports is 256.
+///
+/// In the future, we will use CPU-local storage for each CPU's preemption count,
+/// so we won't need to use a system-wide primitive array here.
+const MAX_CPU_CORES: usize = 2_usize.pow(u8::BITS);
 
 #[allow(clippy::declare_interior_mutable_const)]
 const ATOMIC_U8_ZERO: AtomicU8 = AtomicU8::new(0);
 
 /// The per-core preemption count, indexed by a CPU core's APIC ID.
-/// 
-/// If a CPU's count is `0`, preemption is enabled.
-/// If a CPU's count is greater than `0`, preemption is disabled.
+///
+/// If a CPU's preemption count is `0`, preemption is enabled.
+/// If a CPU's preemption count is greater than `0`, preemption is disabled.
 static PREEMPTION_COUNT: [AtomicU8; MAX_CPU_CORES] = [ATOMIC_U8_ZERO; MAX_CPU_CORES];
 
 
 /// Prevents preemption (preemptive task switching) from occurring
 /// until the returned guard object is dropped.
 pub fn hold_preemption() -> PreemptionGuard {
-    let cpu_id = apic::current_cpu();
-    let prev_val = PREEMPTION_COUNT[cpu_id as usize].fetch_add(1, Ordering::Relaxed);
+    let cpu_id = cpu::current_cpu();
+    let prev_val = PREEMPTION_COUNT[cpu_id.value() as usize].fetch_add(1, Ordering::Relaxed);
```

Comment:
Would you agree if we used the `u8` representation to index into this? that way it would work as-is on most ARMv8 CPUs and panic on other ones. When we have Cpu Local Storage we'd change this.

Reply:
no, that'd require a bunch of extra type wrappers to implement the `Index` trait for `u8`, as it typically is only implemented for `usize`. 
There's no point messing with this, as it'll change soon anyway.

Reply:
If/when you actually encounter CPU ID values larger than `u8::MAX` on aarch64, then we can switch to a map structure like a `BTreeMap`.